### PR TITLE
Bug 1433948 - Part 2 - Add Tracking Protection menu to url long press menu.

### PR DESCRIPTION
**Pull request by farhanpatel**
I've split up the commits to make this easier. I wanted to split up the settings work. But making changes to the user prefs required making the same changes to Settings 😅

Here are the highlights 
- A new way of enabling/disabling TP. It can now for example be turned on in normal mode and turned off in private mode. This is now using 2 userpref bools instead of the old one. I've written a migrator as well.
- A new menu in the location long press. The whitelisting buttons are there but not working. (thats in a separate bug.)
- I've done some general refactoring to ContentBlocker. 
  - move variable declarations and static variables to the top of the class
  - remove redundant *toTab (ex `enableForTab`) a ContentBlocker class is tied to a tab. `tab.contentblocker.enable` is clear.
  - String enum configurations used only by settings are now an enum extension in the settings file.

One thing that is missing is updated tests. 
- This shouldn't break the basic Tracking Protection tests. But any tests that run against the tracking protection settings will break. I'll update those. 
- There are some bugs in the stats not refreshing and whitelisting not working. I'll make sure that the bugs are filed and categorized correctly.

### CODE REVIEW

File: Client/Frontend/Widgets/PhotonActionSheetProtocol.swift [hunk 1]
```diff
@@ -262,5 +262,72 @@ extension PhotonActionSheetProtocol {
...
+            return [tpBlocking]
+            
+        } else if blocker.isEnabled, blocker.isURLWhitelisted(url: currentURL) {
+            // When tracking protection is enabled but the site is in the white list
+            let tpBlocking = PhotonActionSheetItem(title: Strings.SettingsTrackingProtectionSectionName, text: Strings.TrackingProtectionWhiteListOn, iconString: "menu-TrackingProtection", isEnabled: false, accessory: .Disclosure) { _ in
+                guard let bvc = self as? PresentableVC else { return }
+                self.presentSheetWith(title: Strings.SettingsTrackingProtectionSectionName, actions: [[removeFromWhitelist]], on: bvc, from: urlBar)
+            }
+            return [tpBlocking]
+        }
```
[the remark this answers is not in the document]
Comment (farhanpatel) [reply Client/Frontend/Widgets/PhotonActionSheetProtocol.swift]: This was a style choice on my part. I found it makes it easier to follow if each if statement only has one menuItem inside. If there isnt a performance cost of creating menu Items and not using them I think its more clearer to keep them outside of the `if,else` blocks.

File: Client/Frontend/Browser/TabManager.swift [hunk 2]
```diff
@@ -845,7 +845,7 @@ extension TabManager: WKNavigationDelegate {
             let isNoImageMode = self.prefs.boolForKey(PrefsKeys.KeyNoImageModeStatus) ?? false
             tab.noImageMode = isNoImageMode
 
-            if let tpHelper = tab.contentBlocker as? ContentBlockerHelper, !tpHelper.isEnabledForTab {
+            if let tpHelper = tab.contentBlocker as? ContentBlockerHelper, !tpHelper.isEnabled {
```
[the remark this answers is not in the document]
Comment (farhanpatel) [reply Client/Frontend/Browser/TabManager.swift]: I'll file a bug for this. NighMode and NoImadeMode do have static methods that fit a similar pattern. But ContentBlocker does not use a similar pattern.

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 3]
```diff
@@ -199,6 +116,10 @@ class ContentBlockerHelper: NSObject {
         addActiveRulesToTab()
     }
 
+    static func prefsChanged() {
+        NotificationCenter.default.post(name: .ContentBlockerUpdateNeeded, object: nil)
+    }
```
[the remark this answers is not in the document]
Comment (farhanpatel) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Yea. It looks like that. But `prefsChanged` is a static method! So other classes call `ContentBlockerHelper.prefsChanged()` to get all the content blocker prefs to change. TBH the ContentBlockerHelper class mixes too many global responsibilities (parsing/creating block lists) with the tab responsibilities (enabling/disabling for each tab). It makes it pretty hard to follow what is going on. This should be separated but I couldnt in this PR.

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 4]
```diff
@@ -23,147 +23,68 @@ enum BlockList: String {
...
+        static let NormalBrowsingEnabledKey = "prefkey.trackingprotection.normalbrowsing"
+        static let PrivateBrowsingEnabledKey = "prefkey.trackingprotection.privatebrowsing"
     }
 
-    enum PerTabOverrideEnabledState {
-        case notSet
-        case forceEnabledPerTab
-        case forceDisabledPerTab
+    struct Defaults {
+        static let NormalBrowsing = false
```
Comment (justindarc) [inline Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Shouldn't this also be `true` now?

File: Client/Frontend/Widgets/PhotonActionSheet.swift [hunk 5]
```diff
@@ -518,11 +508,17 @@ private class PhotonActionSheetCell: UITableViewCell {
...
-        return siteImageView
+        let statusIcon = UIImageView()
+        statusIcon.contentMode = .scaleAspectFit
+        statusIcon.clipsToBounds = true
+        statusIcon.layer.cornerRadius = PhotonActionSheetCellUX.CornerRadius
+        statusIcon.setContentHuggingPriority(UILayoutPriorityRequired, for: .horizontal)
+        return statusIcon
+    }()
+
+    lazy var disclosureLabel: UILabel = {
```
Comment (justindarc) [inline Client/Frontend/Widgets/PhotonActionSheet.swift]: Any difference between doing this and just `lazy var disclosureLabel = UILabel()` since you're not doing anything else during init?

File: Client/Telemetry/UnifiedTelemetry.swift [hunk 6]
```diff
@@ -28,7 +28,6 @@ class UnifiedTelemetry {
         telemetryConfig.measureUserDefaultsSetting(forKey: "profile.ASPocketStoriesVisible", withDefaultValue: true)
         telemetryConfig.measureUserDefaultsSetting(forKey: "profile.ASBookmarkHighlightsVisible", withDefaultValue: true)
         telemetryConfig.measureUserDefaultsSetting(forKey: "profile.ASRecentHighlightsVisible", withDefaultValue: true)
-        telemetryConfig.measureUserDefaultsSetting(forKey: "profile.prefkey.trackingprotection.enabled", withDefaultValue: "onInPrivateBrowsing")
```
Comment (justindarc) [inline Client/Telemetry/UnifiedTelemetry.swift]: Do we want to measure the two new prefs?

File: Client/Frontend/Widgets/PhotonActionSheet.swift (hunk 5 above)
Comment (farhanpatel) [reply Client/Frontend/Widgets/PhotonActionSheet.swift]: I like to keep the same pattern so when we do decide to add something later we follow it.

File: Client/Telemetry/UnifiedTelemetry.swift (hunk 6 above)
Comment (farhanpatel) [reply Client/Telemetry/UnifiedTelemetry.swift]: Yea. I was going to file a bug for this. There will be 2 now though. One for each browsing mode.

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 7]
```diff
@@ -423,14 +326,73 @@ extension ContentBlockerHelper {
...
+    }
+
+    func whitelist(enable: Bool, forDomain domain: String, completion: (() -> Void)? = nil) {
+        if enable {
+            ContentBlockerHelper.whitelistedDomains.append(domain)
+        } else {
+            ContentBlockerHelper.whitelistedDomains = ContentBlockerHelper.whitelistedDomains.filter { $0 != domain }
+        }
+
+        BlockListChecker.shared.whitelistedDomains = ContentBlockerHelper.whitelistedDomains
```
Comment (garvankeeley) [inline Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Both classes keep a list of whitelistedDomains? Should only be one list no?

Comment (garvankeeley) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Oops, this was in the original code, looks like it is due to the separation of Focus code vs 'our code'. This seemed like a useful distinction when the Focus stats code was added, but I don't think it is anymore -we ended up with various modifications to integrate with our code.
